### PR TITLE
Add minimal structure taken from the hypertriton repository

### PR DESCRIPTION
**Pull request by fmazzasc**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -83,7 +83,8 @@ def __call__(self):
     # List run-time dependencies here. These will be installed by pip when your project is
     # installed. For an analysis of "install_requires" vs pip's requirements files see:
     # https://packaging.python.org/en/latest/requirements.html
-    install_requires=[],
+    install_requires=["uproot==3.10.12", "matplotlib==3.1.1", "pandas==0.25.3", "scikit-learn==0.21.3", "xgboost==0.90",
```
Comment (mpuccio) [inline setup.py]: Should we ask for >= instead of ==?

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+matplotlib==3.1.1
```
Comment (mpuccio) [inline requirements.txt]: Shall we require minimum versions instead of exact versions?

**Comment by mconcas** [comment]
Would be nice, if it is possible already at this state, to add one or more minimal use cases to be used as  tests. If this addition is already able to do something, ofc.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                 c='r', label='Background pdf Test Set')
+
+    hist, bins = np.histogram(
+        prediction[3], bins=bins, range=low_high, density=True)
+    scale = len(prediction[2]) / sum(hist)
+    err = np.sqrt(hist * scale) / scale
+    plt.errorbar(center, hist, yerr=err, fmt='o',
+                 c='b', label='Signal pdf Test Set')
+
+    # plt.gcf().subplots_adjust(left=0.14)
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Remove the commented line

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+    Draw histogram of the DataFrame's series comparing the distribution
+    in `data1` to `data2`.
+
+    Input
+    -----------------------------------------
+    sig_df: signal candidates dataframe
+    bkg_df: background candidates dataframe
+    column: training columns
+    log: log scale plot
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Not all the function inputs are described. Column is not related to the training columns.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+    Output
+    -----------------------------------------
+    res: matplotlib object with the distributions of the features for
+    signal and background
+
+    """
+
+    data1 = bkg_df
+    data2 = sig_df
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Why do you do this?

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    plt.ylabel(r'                                Counts (arb. units)', fontsize=13)
+    plt.legend(frameon=False, fontsize=12)
+    return res
+
+
+def plot_distr(sig_df, bkg_df, column=None, figsize=None, bins=50, log=False):
+    """Build a DataFrame and create two dataset for signal and bkg
+
+    Draw histogram of the DataFrame's series comparing the distribution
+    in `data1` to `data2`.
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: This description is not related to what the function do.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+
+def plot_corr(sig_df, bkg_df, columns, **kwds):
+    """Calculate pairwise correlation between features.
+
+    Input
+    -----------------------------------------------
+    sig_df: signal candidates dataframe
+    bkg_df: background candidates dataframe
+    column: training columns
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: The name is different from the variable and also the description is not correct.

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            self, mc_file_name, data_file_name, sig_selection=0, bkg_selection=0,
+            cent_class=((0, 10),
+                        (10, 30),
+                        (30, 50),
+                        (50, 90))):
+
+        self.cent_class = cent_class.copy()
+        self.n_events = [0, 0, 0, 0]
+
+        self.df_signal = uproot.open(mc_file_name)['SignalTable'].pandas.df()
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: It would be nice to have the name of the trees as data members

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+    corrmat_sig = data_sig.corr(**kwds)
+    corrmat_bkg = data_bkg.corr(**kwds)
+
+    tit = r'$\mathrm{\ \ \ ALICE \ Simulation}$ Pb-Pb $\sqrt{s_{\mathrm{NN}}}$ = 5.02 TeV'
+    fig = plt.figure(figsize=(20, 10))
+    # plt.title(t,y=1.08,fontsize=16)
+    plt.suptitle(tit, fontsize=18, ha='center')
+    grid = ImageGrid(fig, 111, nrows_ncols=(1, 2), axes_pad=0.15, share_all=True,
+                     cbar_location='right', cbar_mode='single', cbar_size='7%', cbar_pad=0.15)
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Maybe this part is relevant only for the hyper-triton analysis and should be reworked.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+    ax1.cax.toggle_label(True)
+    return fig
+
+
+def plot_bdt_eff(threshold, eff_sig):
+    """Plot the BDT efficiency calculated with the function bdt_efficiency_array() in training_utils.py
+
+    Input
+    -----------------------------------
+    threshold: score threscold array
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Typo: threscold

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        self.df_generated = uproot.open(mc_file_name)['GenTable'].pandas.df()
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: If possible I would avoid to have the dataframe for the generated particles as mandatory, since there could be cases in which it is not needed

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        test_set: test set dataframe
+        y_test: test set label array
+
+
+        '''
+        data_range_bkg = '{}<ct<{} and {}<HypCandPt<{} and {}<centrality<{}'.format(
+            ct_range[0], ct_range[1], pt_range[0], pt_range[1], cent_class[0], cent_class[1])
+
+        columns = training_columns.copy()
+        columns.append('InvMass')
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: also here I would avoid hard coded column names

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+    return res
+
+
+def plot_roc(y_truth, model_decision):
+    """Calculate and plot the roc curve
+
+    Input
+    -------------------------------------
+    y_truth: test set label
+    model decision: predicted score
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: What type they should be? Also it looks to me that here the model prediction should be a probability.

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        Output
+        ------------------------------------------------------------
+        train_set: training set dataframe
+        y_train: training set label array
+        test_set: test set dataframe
+        y_test: test set label array
+
+
+        '''
+        data_range_bkg = '{}<ct<{} and {}<HypCandPt<{} and {}<centrality<{}'.format(
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: I would add here a possibility to apply a selection on the invariant mass in case the bkg is taken from right sign data

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+    y_lab: test set label
+    model: trained model
+
+    Output
+    -------------------------------------------
+    res: matplotlib object with shap feature importance
+
+    """
+    subs_bkg = df_in[y_lab == 0].sample(10000)
+    subs_sig = df_in[y_lab == 1].sample(10000)
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: The number of examples used should not be hard coded.

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
+'''Module containing the class to perform the training and the testing of the models '''
+
+
+import pandas as pd
+import uproot
+from sklearn.model_selection import train_test_split
+
+
+class GeneralizedAnalysis:
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: I would generalise the class such that the variables (pT, cT, centrality) are passed as one or more given column of the data frames

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+    return res
+
+
+def plot_precision_recall(y_test, y_score):
+    """ Plot precision recall curve
+
+    Input
+    -------------------------------------
+    y_test: test set label
+    y_score: predicted score
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: What type they should be? Also it looks to me that here y_score should be a probability.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+
+
+def plot_feature_imp(df_in, y_lab, model):
+    """Calculate the feature importance using the shap violin plot for each feature. The calculation
+    is performed on a subsample of the input training/test set. The model could be sklearn or xgboost
+
+    Input
+    -------------------------------------------
+    df_in: test set dataframe
+    y_lab: test set label
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: They do not necessary comes from the test set.

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+    ----------------------------------------
+    res: matplotlib object with the BDT output
+
+
+    '''
+
+    prediction = []
+    for xxx, yyy in ((x_train, y_train), (x_test, y_test)):
+        if model == 'xgb':
+            df1 = clf.predict(xxx[yyy > 0.5][features], output_margin=True)
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: maybe we could add the possibility to use the predicted probability as well

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+
+    """
+
+    data_sig = sig_df[columns]
+    data_bkg = bkg_df[columns]
+
+    corrmat_sig = data_sig.corr(**kwds)
+    corrmat_bkg = data_bkg.corr(**kwds)
+
+    tit = r'$\mathrm{\ \ \ ALICE \ Simulation}$ Pb-Pb $\sqrt{s_{\mathrm{NN}}}$ = 5.02 TeV'
```
Comment (fgrosa) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Remove ALICE simulation and label of the system

File: hipe4ml/TrainingAndTesting/plot_utils.py [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+    precision, recall, _ = precision_recall_curve(y_test, y_score)
+
+    # In matplotlib < 1.5, plt.fill_between does not have a 'step' argument
+    res = plt.figure()
+    step_kwargs = ({'step': 'post'}
+                   if 'step' in signature(plt.fill_between).parameters
+                   else {})
+    plt.step(recall, precision, color='b', alpha=0.2,
+             where='post')
+    plt.fill_between(recall, precision, alpha=0.2, color='b', **step_kwargs)
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/plot_utils.py]: Can be reworked. We are using matplotlib > 3

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+        # dataframe for signal and background with preselection
+        if isinstance(sig_selection, str):
+            self.df_signal = self.df_signal.query(sig_selection)
+        if isinstance(bkg_selection, str):
+            self.df_data_bkg = self.df_data_bkg.query(bkg_selection)
+
+        self.hist_centrality = uproot.open(data_file_name)['EventCounter']
+        self.n_events = []
+        for cent in self.cent_class:
+            self.n_events.append(sum(self.hist_centrality[cent[0]+1:cent[1]]))
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: This part is very dependent on the single analyses. We should think how to make this part easily configurable.

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+
+    def prepare_dataframe(
+            self, training_columns, cent_class=(0, 90), pt_range=(0, 10),
+            ct_range=(0, 100),
+            test=False):
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: Also this function is analysis dependent. It should be reworked to be configurable, as an example the preparation could be done taking a string for the query, user defined, to be more flexible.

File: hipe4ml/TrainingAndTesting/generalized_analysis.py [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+        print('\nnumber of background candidates: {}'.format(len(bkg)))
+        print('number of signal candidates: {}\n'.format(len(sig)))
+
+        df_tot = pd.concat([sig, bkg])
+        train_set, test_set, y_train, y_test = train_test_split(
+            df_tot[columns], df_tot['y'], test_size=0.5, random_state=42)
+
+        return train_set, y_train, test_set, y_test
+
+    def preselection_efficiency(self, cent_class=(0, 90), pt_range=(0, 10), ct_range=(0, 100)):
```
Comment (fcatalan92) [inline hipe4ml/TrainingAndTesting/generalized_analysis.py]: Also this function is analysis dependent and should be more abstracted.

**Comment by fmazzasc** [comment]
Hi @mconcas  @fcatalan92 and @fgrosa, thanks for the comments. I solved most of them and I think that the remaining ones should be discussed together in a further step. For what concerns the tests suggested by @mconcas, me and @mpuccio will work on them in the next days

File: hipe4ml/plot_utils.py [hunk 25]
```diff
@@ -1,45 +1,61 @@
...
+
+    y_train: list or array
+    Training set labels
+
+    x_test: Pandas dataframe
+    Test set
+
+    y_test: list or np.array
+    Test set labels
+
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: This parameters are no more present, right? Moreover, the parameters ```clf```, ```bins``` and  ```raw``` are not described.

File: hipe4ml/plot_utils.py [hunk 26]
```diff
@@ -1,45 +1,61 @@
...
+from pandas.core.index import Index
+import shap
+from mpl_toolkits.axes_grid1 import ImageGrid
 from sklearn.metrics import (auc, average_precision_score,
                              precision_recall_curve, roc_curve)
 
-import shap
-
 
 # plot the BDT score distribution in the train and in the test set for both signal and background
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: Remove this.

File: hipe4ml/plot_utils.py [hunk 27]
```diff
@@ -85,14 +100,22 @@ def plot_distr(sig_df, bkg_df, column=None, figsize=None, bins=50, log=False):
...
+
+    bkg_df: Pandas dataframe
+    background candidates dataframe
+
+    column: list
+    Contains the name of the features you want to plot
+    Example: ['dEdx', 'pT', 'ct']
+
+    log: Bool
+    If True enable log scale plot
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: ```figsize``` and  ```bins``` are not described.

File: hipe4ml/plot_utils.py [hunk 28]
```diff
@@ -148,10 +173,8 @@ def plot_corr(sig_df, bkg_df, columns, **kwds):
     corrmat_sig = data_sig.corr(**kwds)
     corrmat_bkg = data_bkg.corr(**kwds)
 
-    tit = r'$\mathrm{\ \ \ ALICE \ Simulation}$ Pb-Pb $\sqrt{s_{\mathrm{NN}}}$ = 5.02 TeV'
     fig = plt.figure(figsize=(20, 10))
     # plt.title(t,y=1.08,fontsize=16)
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: Remove the commented line

File: hipe4ml/plot_utils.py [hunk 29]
```diff
@@ -226,8 +254,18 @@ def plot_roc(y_truth, model_decision):
...
+    True binary labels. If labels are not either
+    {-1, 1} or {0, 1}, then pos_label should be
+    explicitly given.
+
+    y_score : array
+    Target scores, can either be probability estimates
+    of the positive class, confidence values, or
+    non-thresholded measure of decisions (as returned
+    by “decision_function” on some classifiers).
+
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: The names are different from the function parameters.

File: hipe4ml/plot_utils.py [hunk 30]
```diff
@@ -226,8 +254,18 @@ def plot_roc(y_truth, model_decision):
...
     Input
     -------------------------------------
-    y_truth: test set label
-    model decision: predicted score
+
+    y_true : array
+    True binary labels. If labels are not either
+    {-1, 1} or {0, 1}, then pos_label should be
+    explicitly given.
+
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: What "then pos_label should be explicitly given" means/imply?

File: hipe4ml/plot_utils.py [hunk 31]
```diff
@@ -280,24 +325,26 @@ def plot_precision_recall(y_test, y_score):
 
     Input
     -------------------------------------
-    y_test: test set label
-    y_score: predicted score
+    y_test: array
+    True binary labels. If labels are not either
+    {-1, 1} or {0, 1}, then pos_label should be
+    explicitly given.
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: What "then pos_label should be explicitly given" means/imply?

File: hipe4ml/training_utils.py [hunk 32]
```diff
@@ -7,69 +7,105 @@
...
+    nfold: int
+    Number of folds to calculate the cross
+    validation error
 
-    nfold: number of folds to calculate the cross validation error
+    metrics: string, callable or None
+    A string (see model evaluation documentation) or a
+    scorer callable object / function with signature
+    scorer(estimator, X, y) which should return only a
+    single value.
```
Comment (fcatalan92) [inline hipe4ml/training_utils.py]: Maybe here a link to the scikit-learn documentation is needed.

File: hipe4ml/training_utils.py [hunk 33]
```diff
@@ -7,69 +7,105 @@
...
 
 
-def evaluate_hyperparams(data, training_columns, reg_params, hyp_params, nfold=5):
-    '''
+def evaluate_hyperparams(data, training_columns, reg_params, hyp_params, metrics, nfold=5):
+    """
     Calculate for a set of hyperparams the cross val score
 
     Input
     --------------------------------------------------------
```
Comment (fcatalan92) [inline hipe4ml/training_utils.py]: The explanation of  ```training_columns``` is missing.

File: hipe4ml/training_utils.py [hunk 34]
```diff
@@ -7,69 +7,105 @@
...
 
-def optimize_params_bayes(data, training_columns, reg_params, hyperparams_ranges, nfold=3, init_points=5,
+def optimize_params_bayes(data, training_columns, reg_params, hyperparams_ranges, metrics, nfold=5, init_points=5,
                           n_iter=5):
-    '''
+    """
     Perform Bayesian optimization
 
     Input
     ------------------------------------------------------
```
Comment (fcatalan92) [inline hipe4ml/training_utils.py]: The explanation of ```training_columns``` is missing.

File: hipe4ml/plot_utils.py [hunk 35]
```diff
@@ -9,8 +9,6 @@
 from sklearn.metrics import (auc, average_precision_score,
                              precision_recall_curve, roc_curve)
 
-
-# plot the BDT score distribution in the train and in the test set for both signal and background
 def plot_output_train_test(
         clf, data, model='xgb', features=None, bins=80, raw=True, **kwds):
     """
```
Comment (fcatalan92) [inline hipe4ml/plot_utils.py]: The description of ```clf``` is still missing.
